docs/design/resource-qos.md

```diff
@@ -0,0 +1,220 @@
+- Containers will be throttled if they exceed their limit. If limit is unspecified, then the containers can use excess CPU when available.
+
+### Incompressible Resource Guarantees
+
+- For now, we are only supporting memory.
+- Containers will get the amount of memory they request, if they exceed their memory request, they could be killed (if some other container needs memory), but if containers consume fewer resources than requested, they will not be killed (except in cases where system tasks or daemons need more memory).
+- Containers will be killed if they use more memory than their limit.
+
+### Kubelet Admission Policy
+
+- Pods will be admitted by Kubelet based on the sum of requests of its containers. The Kubelet will ensure that sum of requests of all containers (over all pods) is within the system’s resource (for both memory and CPU).
+
+## QoS Classes
+
+In an overcommitted system (where sum of requests > machine capacity) containers might eventually have to be killed, for example if the system runs out of CPU or memory resources. Ideally, we should kill containers that are less important. For each resource, we divide containers into 3 QoS classes: *Guaranteed*, *Burstable*, and *Best-Effort*, in decreasing order of priority.
+
+The relationship between "Requests and Limits" and "QoS Classes" is subtle. Theoretically, the policy of classifying containers into QoS classes is orthogonal to the requests and limits specified for the container. Hypothetically, users could use an (currently unplanned) API to specify whether a container is guaranteed or best-effort. However, in the current design, the policy of classifying containers into QoS classes is intimately tied to "Requests and Limits" - in fact, QoS classes are used to implement some of the memory guarantees described in the previous section.
+
+Containers can be of one of 3 different classes:
+
+- If `limits` and optionally `requests` are set (not equal to `0`) for all resources and they are *equal* then the container is classified as **Guaranteed**.
+
+Examples:
+
+```json
+resources:
+  limits:
+    cpu: 10m
+    memory: 1Gi
+```
+
+```json
+resources:
+  limits:
+    cpu: 10m
+    memory: 1Gi
+  requests:
+    cpu: 10m
+    memory: 1Gi
+```
+
+- If `requests` and optionally `limits` are set (not equal to `0`) for all resources and they are *not equal* then the container is classified as **Burstable**.
+When limits not specified, they default to the node capacity.
+
+Examples:
+
+```json
+resources:
+  limits:
+    cpu: 1
+    memory: 10Gi
+  requests:
+    memory: 1Gi
+```
+
+```json
+resources:
+  requests:
+    cpu: 10m
+    memory: 1Gi
+```
+
+- If `requests` and `limits` are not set for any of the resources then the container is classified as **Best-Effort**.
+When limits not specified, they default to the node capacity.
```

Comment:
@davidopp - it takes the docker default behavior today which is machine capacity.